### PR TITLE
Add rest of MsQuic transport.

### PR DESCRIPTION
**Pull request by jkotalik**
This adds a MsQuic transport with a sample client and server. There are many things to clean up here, but we don't know if we are going to keep this for long anyway as the System.Net.Quic will supersede this.

### CODE REVIEW

File: src/Servers/Kestrel/Transport.MsQuic/src/Internal/MsQuicNativeMethods.cs [hunk 1]
```diff
@@ -470,6 +471,114 @@ internal unsafe struct QuicBuffer
...
+        private const ushort AF_INET6 = 23;
+        private const ushort AF_UNSPEC = 0;
+
+        private static SOCKADDR_INET ConvertToSocketAddrInet(IPAddress ipAddress)
+        {
+            var lResult = new SOCKADDR_INET();
+            switch (ipAddress.AddressFamily)
+            {
+                case AddressFamily.InterNetwork:
+                    lResult.Ipv4.sin_addr0 = ipAddress.GetAddressBytes()[0];
```
Comment (davidfowl) [inline src/Servers/Kestrel/Transport.MsQuic/src/Internal/MsQuicNativeMethods.cs]: Doesn't this allocate a crazy number of arrays?

File: src/Servers/Kestrel/Transport.MsQuic/src/Internal/MsQuicStream.cs [hunk 2]
```diff
@@ -0,0 +1,241 @@
...
+    internal class MsQuicStream : TransportConnection
+    {
+        private Task _processingTask;
+        private QuicStream _stream;
+        private MsQuicConnection _connection;
+        private readonly CancellationTokenSource _streamClosedTokenSource = new CancellationTokenSource();
+
+        public MsQuicStream(MsQuicConnection connection, QUIC_STREAM_OPEN_FLAG flags)
+        {
+            var inputOptions = new PipeOptions(MemoryPool, PipeScheduler.ThreadPool, PipeScheduler.Inline, 1024 * 1024, 1024 * 1024 / 2, useSynchronizationContext: false);
```
Comment (davidfowl) [inline src/Servers/Kestrel/Transport.MsQuic/src/Internal/MsQuicStream.cs]: Make those numbers constants in here.

File: src/Servers/Kestrel/Transport.MsQuic/src/Internal/MsQuicStream.cs [hunk 3]
```diff
@@ -0,0 +1,241 @@
...
+            var outputOptions = new PipeOptions(MemoryPool, PipeScheduler.Inline, PipeScheduler.ThreadPool, 1024 * 64, 1024 * 64 / 2, useSynchronizationContext: false);
+
+            var pair = DuplexPipe.CreateConnectionPair(inputOptions, outputOptions);
+            if (flags.HasFlag(QUIC_STREAM_OPEN_FLAG.UNIDIRECTIONAL))
+            {
+                Features.Set<IUnidirectionalStreamFeature>(new UnidirectionalStreamFeature());
+            }
+
+            // Add a fake TLS connection feature becuase quic is always secure
+            var feature = new FakeTlsConnectionFeature();
```
Comment (davidfowl) [inline src/Servers/Kestrel/Transport.MsQuic/src/Internal/MsQuicStream.cs]: Shouldn't this be real and expose the actual TLS information?

File: src/Servers/Kestrel/Transport.MsQuic/src/Internal/MsQuicNativeMethods.cs (hunk 1 above)
Comment (jkotalik) [reply src/Servers/Kestrel/Transport.MsQuic/src/Internal/MsQuicNativeMethods.cs]: I need to play around with PhysicalAddress here.

File: src/Servers/Kestrel/Transport.MsQuic/src/Internal/MsQuicStream.cs [hunk 4]
```diff
@@ -0,0 +1,241 @@
...
+            var output = Output;
+            try
+            {
+                while (true)
+                {
+                    var result = await output.ReadAsync();
+
+                    if (result.IsCanceled)
+                    {
+                        break;
```
Comment (davidfowl) [inline src/Servers/Kestrel/Transport.MsQuic/src/Internal/MsQuicStream.cs]: Is this graceful of abort?

File: src/Servers/Kestrel/Transport.MsQuic/src/Internal/MsQuicStream.cs [hunk 5]
```diff
@@ -0,0 +1,241 @@
...
+        private uint HandleEventSendShutdownComplete(ref MsQuicNativeMethods.StreamEvent evt)
+        {
+            return MsQuicConstants.Success;
+        }
+
+        private uint HandleEventPeerSendClose()
+        {
+            // TODO complete async
+            // Close as fast as possible here.
+            Input.Complete();
```
Comment (davidfowl) [inline src/Servers/Kestrel/Transport.MsQuic/src/Internal/MsQuicStream.cs]: Does this happen after *all* reads occur? That is will HandleEventRecv ever be called after this?

File: src/Servers/Kestrel/Transport.MsQuic/src/Internal/MsQuicTransportFactory.cs [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+
+            var logger = loggerFactory.CreateLogger("Microsoft.AspNetCore.Server.Kestrel.Transport.MsQuic");
+            var trace = new MsQuicTrace(logger);
+
+            _transportContext = new MsQuicTransportContext(applicationLifetime, trace, options.Value);
+        }
+
+        public async ValueTask<IConnectionListener> BindAsync(EndPoint endpoint, CancellationToken cancellationToken = default)
+        {
+
```
Comment (davidfowl) [inline src/Servers/Kestrel/Transport.MsQuic/src/Internal/MsQuicTransportFactory.cs]: nit: Space

File: src/Servers/Connections.Abstractions/src/Features/IStreamListener.cs [hunk 7]
```diff
@@ -0,0 +1,14 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System.Threading.Tasks;
+
+namespace Microsoft.AspNetCore.Connections.Features
+{
+    public interface IStreamListener
```
Comment (davidfowl) [inline src/Servers/Connections.Abstractions/src/Features/IStreamListener.cs]: I wonder if these should be Quic specific features?

File: src/Servers/Connections.Abstractions/src/Features/IStreamListener.cs [hunk 8]
```diff
@@ -0,0 +1,14 @@
...
+    {
+        ValueTask<ConnectionContext> StartUnidirectionalStreamAsync();
```
Comment (davidfowl) [inline src/Servers/Connections.Abstractions/src/Features/IStreamListener.cs]: Usually we have minimal features (like a method per interface) or something grouping related methods together. For example, a multiplexed connection would have `AcceptAsync`. The `StartBidirectionalStreamAsync` and `StartUnidirectionalStreamAsync` seem unrelated

File: src/Servers/Kestrel/Transport.MsQuic/src/Internal/MsQuicConnection.cs [hunk 9]
```diff
@@ -0,0 +1,248 @@
...
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Connections;
+using Microsoft.AspNetCore.Connections.Features;
+
+namespace Microsoft.AspNetCore.Server.Kestrel.Transport.MsQuic.Internal
+{
+    internal class MsQuicConnection : TransportConnection, IStreamListener
+    {
+        private QuicConnection _connection;
+        private IAsyncEnumerator<MsQuicStream> _acceptEnumerator;
```
Comment (davidfowl) [inline src/Servers/Kestrel/Transport.MsQuic/src/Internal/MsQuicConnection.cs]: Get rid of this. It's not needed here, just inline the accept code into the AcceptAsync method itself.

File: src/Servers/Kestrel/Transport.MsQuic/src/Internal/MsQuicConnection.cs [hunk 10]
```diff
@@ -0,0 +1,248 @@
...
+
+            return null;
+        }
+
+        public async ValueTask<ConnectionContext> StartBidirectionalStreamAsync()
+        {
+            var flags = QUIC_STREAM_OPEN_FLAG.NONE;
+            var msquicStream = new MsQuicStream(this, flags);
+            var stream = _connection.StreamOpen(flags, msquicStream.HandleStreamEvent);
+            await stream.StartAsync(QUIC_STREAM_START_FLAG.NONE);
```
Comment (davidfowl) [inline src/Servers/Kestrel/Transport.MsQuic/src/Internal/MsQuicConnection.cs]: pass the flasg here too?

File: src/Servers/Kestrel/samples/QuicSampleClient/Program.cs [hunk 11]
```diff
@@ -0,0 +1,58 @@
...
+{
+    class Program
+    {
+        static async Task Main(string[] args)
+        {
+            var transportOptions = new MsQuicTransportOptions()
+            {
+                Alpn = "QuicTest",
+                RegistrationName = "Quic-AspNetCore-client",
+                Certificate = CertificateLoader.LoadFromStoreCert("localhost", StoreName.My.ToString(), StoreLocation.CurrentUser, true),
```
Comment (davidfowl) [inline src/Servers/Kestrel/samples/QuicSampleClient/Program.cs]: Are you going to check in a cert?

File: src/Servers/Kestrel/samples/QuicSampleClient/Program.cs [hunk 12]
```diff
@@ -0,0 +1,58 @@
...
+
+            var transportContext = new MsQuicTransportContext(null, null, transportOptions);
+            var factory = new MsQuicConnectionFactory(transportContext);
+            var connectionContext = await factory.ConnectAsync(new IPEndPoint(IPAddress.Loopback, 5555));
+            var streamFeature = connectionContext.Features.Get<IStreamListener>();
+            var streamContext = await streamFeature.StartBidirectionalStreamAsync();
+
+            Console.CancelKeyPress += new ConsoleCancelEventHandler((sender, args) =>
+            {
+                streamContext.Transport.Output.Complete();
```
Comment (davidfowl) [inline src/Servers/Kestrel/samples/QuicSampleClient/Program.cs]: Complete after the loop and wait on the Console like this https://github.com/davidfowl/BedrockTransports/blob/4f2bc78b14e69fb7420919cdedab300a9ec6a3ea/ServerApplication/Program.cs#L73-L75

File: src/Servers/Connections.Abstractions/src/Features/IStreamListener.cs (hunk 7 above)
Comment (jkotalik) [reply src/Servers/Connections.Abstractions/src/Features/IStreamListener.cs]: I think they should for now. Rename to IQuicStreamListener?

File: src/Servers/Connections.Abstractions/src/Features/IStreamListener.cs (hunk 8 above)
Comment (jkotalik) [reply src/Servers/Connections.Abstractions/src/Features/IStreamListener.cs]: Yeah, I think I can break this up into 3 features: IQuicStreamListener, IStartStreamFeature (or some name), and some feature for knowing if the stream is unidirectional or bidirectional.

File: src/Servers/Connections.Abstractions/src/Features/IUnidirectionalStreamFeature.cs [hunk 13]
```diff
@@ -0,0 +1,9 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+namespace Microsoft.AspNetCore.Connections.Features
+{
+    public interface IUnidirectionalStreamFeature
```
Comment (jkotalik) [inline src/Servers/Connections.Abstractions/src/Features/IUnidirectionalStreamFeature.cs]: How bad would it be to detect unidirectional via checking the output pipe being null or not?

File: src/Servers/Kestrel/Transport.MsQuic/src/Internal/MsQuicStream.cs (hunk 3 above)
Comment (jkotalik) [reply src/Servers/Kestrel/Transport.MsQuic/src/Internal/MsQuicStream.cs]: I'll leave a TODO/file an issue for client certs.

File: src/Servers/Kestrel/Transport.MsQuic/src/Internal/MsQuicStream.cs (hunk 4 above)
Comment (jkotalik) [reply src/Servers/Kestrel/Transport.MsQuic/src/Internal/MsQuicStream.cs]: I don't think it can be canceled here.

Comment (jkotalik) [reply src/Servers/Kestrel/Transport.MsQuic/src/Internal/MsQuicStream.cs]: Actually, if I implement abort, it would only for abort. I'll do that now.

File: src/Servers/Kestrel/Transport.MsQuic/src/Internal/MsQuicStream.cs (hunk 5 above)
Comment (jkotalik) [reply src/Servers/Kestrel/Transport.MsQuic/src/Internal/MsQuicStream.cs]: I think that's an invariant but can confirm.

File: src/Servers/Kestrel/samples/QuicSampleClient/Program.cs (hunk 11 above)
Comment (jkotalik) [reply src/Servers/Kestrel/samples/QuicSampleClient/Program.cs]: This is the dev cert usually.

File: src/Servers/Kestrel/samples/QuicSampleClient/Program.cs (hunk 12 above)
Comment (jkotalik) [reply src/Servers/Kestrel/samples/QuicSampleClient/Program.cs]: Still has the same issue of calling Complete in the middle of the loop. How do you cause the loop to exit here normally? Use CancelPendingRead?

File: src/Servers/Connections.Abstractions/src/Features/IUnidirectionalStreamFeature.cs (hunk 13 above)
Comment (Tratcher) [reply src/Servers/Connections.Abstractions/src/Features/IUnidirectionalStreamFeature.cs]: Terrible 😁. At least have a bool or enum property.

File: src/Servers/Kestrel/Transport.MsQuic/src/Internal/MsQuicNativeMethods.cs [hunk 14]
```diff
@@ -536,5 +494,105 @@ internal struct SOCKADDR_INET
...
+            var socketAddress = new SOCKADDR_INET();
+            var buffer = endpoint.Address.GetAddressBytes();
+            switch (endpoint.Address.AddressFamily)
+            {
+                case AddressFamily.InterNetwork:
+                    socketAddress.Ipv4.sin_addr0 = buffer[0];
+                    socketAddress.Ipv4.sin_addr1 = buffer[1];
+                    socketAddress.Ipv4.sin_addr2 = buffer[2];
+                    socketAddress.Ipv4.sin_addr3 = buffer[3];
+                    socketAddress.Ipv4.sin_family = 0;
```
Comment (Tratcher) [inline src/Servers/Kestrel/Transport.MsQuic/src/Internal/MsQuicNativeMethods.cs]: Shouldn't this be set?

Comment (jkotalik) [reply src/Servers/Kestrel/Transport.MsQuic/src/Internal/MsQuicNativeMethods.cs]: It was set to zero before. This means that MsQuic will listen for both IPv4 and IPv6 for the same port. Not sure if we should allow this to be configurable or not.

File: src/Servers/Kestrel/Transport.MsQuic/src/Internal/UnidirectionalStreamFeature.cs [hunk 15]
```diff
@@ -0,0 +1,12 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using Microsoft.AspNetCore.Connections.Features;
+
+namespace Microsoft.AspNetCore.Server.Kestrel.Transport.MsQuic.Internal
+{
+    internal class UnidirectionalStreamFeature : IUnidirectionalStreamFeature
```
Comment (halter73) [inline src/Servers/Kestrel/Transport.MsQuic/src/Internal/UnidirectionalStreamFeature.cs]: How/where will this be used?

File: src/Servers/Kestrel/Transport.MsQuic/src/Internal/SendContext.cs [hunk 16]
```diff
@@ -0,0 +1,78 @@
...
+
+using System;
+using System.Buffers;
+using System.Runtime.InteropServices;
+using System.Threading.Tasks;
+using static Microsoft.AspNetCore.Server.Kestrel.Transport.MsQuic.Internal.MsQuicNativeMethods;
+
+namespace Microsoft.AspNetCore.Server.Kestrel.Transport.MsQuic.Internal
+{
+    public sealed class SendContext
```
Comment (halter73) [inline src/Servers/Kestrel/Transport.MsQuic/src/Internal/SendContext.cs]: Eventually we want this to be a SafeHandle, right?

File: src/Servers/Kestrel/Transport.MsQuic/src/Internal/MsQuicConstants.cs [hunk 17]
```diff
@@ -8,7 +8,11 @@ namespace Microsoft.AspNetCore.Server.Kestrel.Transport.MsQuic.Internal
 {
     internal static class MsQuicConstants
     {
-        private const uint Success = 0;
+        public const uint SuccessConst = 0;
+
+        internal static uint InternalError = RuntimeInformation.IsOSPlatform(OSPlatform.Windows) ? Windows.InternalError : Linux.InternalError;
+        internal static uint Success = 0;
```
Comment (halter73) [inline src/Servers/Kestrel/Transport.MsQuic/src/Internal/MsQuicConstants.cs]: Why have Success and SuccesConst when they're both the same type and value? Why make SuccesConst public if it's only referenced from a nested class?

File: src/Servers/Kestrel/Transport.MsQuic/src/Internal/MsQuicConnection.cs [hunk 18]
```diff
@@ -0,0 +1,425 @@
...
+
+        private readonly Channel<MsQuicStream> _acceptQueue = Channel.CreateUnbounded<MsQuicStream>(new UnboundedChannelOptions
+        {
+            SingleReader = true,
+            SingleWriter = true
+        });
+
+        public MsQuicConnection(MsQuicApi registration, MsQuicTransportContext context, IntPtr nativeObjPtr)
+        {
+            Registration = registration;
```
Comment (nibanks) [inline src/Servers/Kestrel/Transport.MsQuic/src/Internal/MsQuicConnection.cs]: Rename `[Rr]egistration` to `Api`, or something along those lines.

File: src/Servers/Kestrel/Transport.MsQuic/src/Internal/MsQuicConnection.cs [hunk 19]
```diff
@@ -0,0 +1,425 @@
...
+            _context = context;
+            _nativeObjPtr = nativeObjPtr;
+
+            ConnectionCallbackDelegate nativeCallback = NativeCallbackHandler;
+            _unmanagedFnPtrForNativeCallback = Marshal.GetFunctionPointerForDelegate(nativeCallback);
+            SetCallbackHandler(HandleEvent);
+            SetIdleTimeout(TimeSpan.FromMinutes(10));
```
Comment (nibanks) [inline src/Servers/Kestrel/Transport.MsQuic/src/Internal/MsQuicConnection.cs]: I'd prefer you let the transport default stand, unless you have a value from an upper layer to overwrite it.

File: src/Servers/Kestrel/Transport.MsQuic/src/Internal/MsQuicConnection.cs [hunk 20]
```diff
@@ -0,0 +1,425 @@
...
+        public uint HandleEvent(
+            ref MsQuicNativeMethods.ConnectionEvent evt)
+        {
+            var status = HandleEventCore(ref evt);
+
+            return status;
+        }
+
+        private uint HandleEventCore(
+          ref MsQuicNativeMethods.ConnectionEvent connectionEvent)
```
Comment (nibanks) [inline src/Servers/Kestrel/Transport.MsQuic/src/Internal/MsQuicConnection.cs]: Why not combine these functions?

File: src/Servers/Kestrel/Transport.MsQuic/src/Internal/MsQuicNativeMethods.cs (hunk 14 above)
Comment (nibanks) [reply src/Servers/Kestrel/Transport.MsQuic/src/Internal/MsQuicNativeMethods.cs]: You do need to set the family. Yes, msquic supports listening to AF_UNSPEC, but if the endpoint is a v4 or v6 you should set the family accordingly.

File: src/Servers/Kestrel/Transport.MsQuic/src/Internal/MsQuicStream.cs [hunk 21]
```diff
@@ -0,0 +1,413 @@
...
+        private IntPtr _nativeObjPtr;
+        private GCHandle _handle;
+        private StreamCallback _callback;
+        private readonly IntPtr _unmanagedFnPtrForNativeCallback;
+
+        public MsQuicStream(MsQuicApi registration, MsQuicConnection connection, MsQuicTransportContext context, QUIC_STREAM_OPEN_FLAG flags, IntPtr nativeObjPtr)
+        {
+            Debug.Assert(connection != null);
+
+            Registration = registration;
```
Comment (nibanks) [inline src/Servers/Kestrel/Transport.MsQuic/src/Internal/MsQuicStream.cs]: `Registration` to `Api`

File: src/Servers/Kestrel/Transport.MsQuic/src/Internal/MsQuicStream.cs [hunk 22]
```diff
@@ -0,0 +1,413 @@
...
+            var status = MsQuicConstants.Success;
+
+            switch (evt.Type)
+            {
+                case QUIC_STREAM_EVENT.START_COMPLETE:
+                    status = HandleStartComplete();
+                    break;
+                case QUIC_STREAM_EVENT.RECV:
+                    {
+                        status = HandleEventRecv(
```
Comment (nibanks) [inline src/Servers/Kestrel/Transport.MsQuic/src/Internal/MsQuicStream.cs]: Note, you aren't allowed to fail the RECEIVE event.

File: src/Servers/Kestrel/Transport.MsQuic/src/Internal/MsQuicStream.cs (hunk 5 above)
Comment (nibanks) [reply src/Servers/Kestrel/Transport.MsQuic/src/Internal/MsQuicStream.cs]: It's guaranteed to come after all receives.

File: src/Servers/Kestrel/Transport.MsQuic/src/Internal/MsQuicStream.cs [hunk 23]
```diff
@@ -0,0 +1,413 @@
...
+                sendContext.Buffers,
+                sendContext.BufferCount,
+                (uint)flags,
+                sendContext.NativeClientSendContext);
+            MsQuicStatusException.ThrowIfFailed(status);
+
+            return sendContext.TcsTask; // TODO make QuicStream implement IValueTaskSource?
+        }
+
+        public Task StartAsync(QUIC_STREAM_START_FLAG flags)
```
Comment (nibanks) [inline src/Servers/Kestrel/Transport.MsQuic/src/Internal/MsQuicStream.cs]: It's not async unless the appropriate flag is set.

File: src/Servers/Kestrel/Transport.MsQuic/src/Internal/MsQuicStream.cs [hunk 24]
```diff
@@ -0,0 +1,413 @@
...
+            }
+
+            _nativeObjPtr = IntPtr.Zero;
+            Registration = null;
+
+            _disposed = true;
+        }
+    }
+
+    internal class FakeTlsConnectionFeature : ITlsConnectionFeature
```
Comment (nibanks) [inline src/Servers/Kestrel/Transport.MsQuic/src/Internal/MsQuicStream.cs]: What's the point of this?

File: src/Servers/Kestrel/Transport.MsQuic/src/Internal/QuicSession.cs [hunk 25]
```diff
@@ -0,0 +1,143 @@
...
+        private IntPtr _nativeObjPtr;
+        private MsQuicApi _registration;
+
+        internal QuicSession(MsQuicApi registration, IntPtr nativeObjPtr)
+        {
+            _registration = registration;
+            _nativeObjPtr = nativeObjPtr;
+        }
+
+        public async ValueTask<MsQuicConnection> ConnectionOpenAsync(IPEndPoint endpoint, MsQuicTransportContext context)
```
Comment (nibanks) [inline src/Servers/Kestrel/Transport.MsQuic/src/Internal/QuicSession.cs]: Open and Start are explicitly separate in the msquic API for a reason. For the 0-RTT case, you open streams and send on them **before** starting the connection.

File: src/Servers/Kestrel/Transport.MsQuic/src/MsQuicTransportOptions.cs [hunk 26]
```diff
@@ -0,0 +1,49 @@
...
+
+        /// <summary>
+        /// The maximum number of concurrent inbound uni-directional streams per connection.
+        /// </summary>
+        public ushort MaxUnidirectionalStreamCount { get; set; } = 10;
+
+        /// <summary>
+        /// The Application Layer Protocol Negotiation string.
+        /// </summary>
+        public string Alpn { get; set; }
```
Comment (nibanks) [inline src/Servers/Kestrel/Transport.MsQuic/src/MsQuicTransportOptions.cs]: Note that QUIC requires ALPN.

File: src/Servers/Kestrel/Transport.MsQuic/src/Internal/MsQuicNativeMethods.cs (hunk 14 above)
Comment (Tratcher) [reply src/Servers/Kestrel/Transport.MsQuic/src/Internal/MsQuicNativeMethods.cs]: Unspecified only makes sense when you're not setting an IP.

Comment (nibanks) [reply src/Servers/Kestrel/Transport.MsQuic/src/Internal/MsQuicNativeMethods.cs]: MsQuic supports listening on unspecified address, but a specified port.

File: src/Servers/Kestrel/Transport.MsQuic/src/Internal/MsQuicNativeMethods.cs [hunk 27]
```diff
@@ -536,5 +494,105 @@ internal struct SOCKADDR_INET
...
+            {
+                case AddressFamily.InterNetwork:
+                    socketAddrInet.Ipv4.sin_port = convertedPort;
+                    break;
+                case AddressFamily.InterNetworkV6:
+                    socketAddrInet.Ipv6.sin6_port = convertedPort;
+                    break;
+                default:
+                    socketAddrInet.Ipv6.sin6_port = convertedPort;
+                    break;
```
Comment (Tratcher) [inline src/Servers/Kestrel/Transport.MsQuic/src/Internal/MsQuicNativeMethods.cs]: Combine these cases if they're the same.

File: src/Servers/Kestrel/Transport.MsQuic/src/Internal/MsQuicStream.cs [hunk 28]
```diff
@@ -0,0 +1,420 @@
...
+            var outputOptions = new PipeOptions(MemoryPool, PipeScheduler.Inline, PipeScheduler.ThreadPool, maxWriteBufferSize, maxWriteBufferSize / 2, useSynchronizationContext: false);
+
+            var pair = DuplexPipe.CreateConnectionPair(inputOptions, outputOptions);
+
+            if (flags.HasFlag(QUIC_STREAM_OPEN_FLAG.UNIDIRECTIONAL))
+            {
+                Features.Set<IUnidirectionalStreamFeature>(this);
+            }
+
+            // TODO populate the ITlsConnectionFeature (requires client certs).
```
Comment (Tratcher) [inline src/Servers/Kestrel/Transport.MsQuic/src/Internal/MsQuicStream.cs]: We'll also need ITlsHandshakeFeature, though that makes more sense at the connection level.

File: src/Servers/Kestrel/Transport.MsQuic/src/Internal/MsQuicStream.cs [hunk 29]
```diff
@@ -0,0 +1,420 @@
...
+        private Task _processingTask;
+        private MsQuicConnection _connection;
+        private readonly CancellationTokenSource _streamClosedTokenSource = new CancellationTokenSource();
+        private IMsQuicTrace _trace;
+        private bool _disposed;
+        private IntPtr _nativeObjPtr;
+        private GCHandle _handle;
+        private readonly IntPtr _unmanagedFnPtrForNativeCallback;
+
+        internal ResettableCompletionSource _resettableCts;
```
Comment (Tratcher) [inline src/Servers/Kestrel/Transport.MsQuic/src/Internal/MsQuicStream.cs]: Cts = CancellationTokenSource?
Tcs = TaskCompletionSource?
_resettableCompletion?

File: src/Servers/Kestrel/Transport.MsQuic/src/Internal/MsQuicStream.cs (hunk 3 above)
Comment (Tratcher) [reply src/Servers/Kestrel/Transport.MsQuic/src/Internal/MsQuicStream.cs]: You should remove FakeTlsConnectionFeature, ITlsConnectionFeature is optional.

File: src/Servers/Kestrel/Transport.MsQuic/src/MsQuicTransportOptions.cs (hunk 26 above)
Comment (Tratcher) [reply src/Servers/Kestrel/Transport.MsQuic/src/MsQuicTransportOptions.cs]: Shouldn't ALPN be a list?

File: src/Servers/Kestrel/samples/QuicSampleClient/Program.cs (hunk 11 above)
Comment (Tratcher) [reply src/Servers/Kestrel/samples/QuicSampleClient/Program.cs]: The dev cert is OK for samples, just not tests.

File: src/Servers/Kestrel/Transport.MsQuic/src/Internal/MsQuicNativeMethods.cs (hunk 14 above)
Comment (jkotalik) [reply src/Servers/Kestrel/Transport.MsQuic/src/Internal/MsQuicNativeMethods.cs]: > Unspecified only makes sense when you're not setting an IP.

So listening to any address (0.0.0.0) on a certain port?

File: src/Servers/Kestrel/Transport.MsQuic/src/Internal/SendContext.cs (hunk 16 above)
Comment (jkotalik) [reply src/Servers/Kestrel/Transport.MsQuic/src/Internal/SendContext.cs]: Yeah, too big of a change for this PR.

File: src/Servers/Kestrel/Transport.MsQuic/src/Internal/MsQuicNativeMethods.cs (hunk 14 above)
Comment (Tratcher) [reply src/Servers/Kestrel/Transport.MsQuic/src/Internal/MsQuicNativeMethods.cs]: Yeah, I think this code only works right now because the sample is passing in 0.0.0.0. If you put in a specific IP without setting the address family it probably won't work.

File: src/Servers/Kestrel/Transport.MsQuic/src/Internal/MsQuicStream.cs (hunk 28 above)
Comment (jkotalik) [reply src/Servers/Kestrel/Transport.MsQuic/src/Internal/MsQuicStream.cs]: I'll leave a todo in the connection

File: src/Servers/Kestrel/Transport.MsQuic/src/Internal/MsQuicStream.cs (hunk 24 above)
Comment (jkotalik) [reply src/Servers/Kestrel/Transport.MsQuic/src/Internal/MsQuicStream.cs]: Pretty much it's a feature that is checked up stack (Kestrel) to make sure the connection is "secure". Right now, we can't get this information from MsQuic (correct me if I'm wrong), so we have a fake one for now.

File: src/Servers/Kestrel/samples/QuicSampleClient/Program.cs (hunk 12 above)
Comment (halter73) [reply src/Servers/Kestrel/samples/QuicSampleClient/Program.cs]: To gracefully exit even when there's write backpressure, you probably want to look into calling CancelPendingWrite too. Also, does Console.ReadLine() unblock on Ctrl+C? You probably want to look into how to interrupt the ReadLine call and handle the null return value it might produce.

Or maybe, since this is just a sample, we just never complete the PipeWriter and just call Environment.Exit(0) in the CancelKeyPress event handler.

File: src/Servers/Kestrel/Transport.MsQuic/src/MsQuicTransportOptions.cs (hunk 26 above)
Comment (jkotalik) [reply src/Servers/Kestrel/Transport.MsQuic/src/MsQuicTransportOptions.cs]: Follow up PR. @nibanks do we need to create a session per ALPN, or just a ';' separated list.

File: src/Servers/Kestrel/samples/QuicSampleClient/Program.cs (hunk 12 above)
Comment (jkotalik) [reply src/Servers/Kestrel/samples/QuicSampleClient/Program.cs]: I just pushed a change to call CancelPendingRead and CancelPendingWrite.

File: src/Servers/Kestrel/Transport.MsQuic/src/MsQuicTransportOptions.cs (hunk 26 above)
Comment (nibanks) [reply src/Servers/Kestrel/Transport.MsQuic/src/MsQuicTransportOptions.cs]: Currently, MsQuic doesn't support multiple ALPNs for the same session, but it's been on my wish list for some time. It would probably be best to take a `;` separated list (string) to be passed into the Session. It's not high on my priority list right now, but feel free to make a PR to MsQuic if you feel like it.

File: src/Servers/Kestrel/Transport.MsQuic/src/IMsQuicTrace.cs [hunk 30]
```diff
@@ -1,12 +1,16 @@
 // Copyright (c) .NET Foundation. All rights reserved.
 // Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
 
+using System;
 using Microsoft.Extensions.Logging;
 
 namespace Microsoft.AspNetCore.Server.Kestrel.Transport.MsQuic
 {
     public interface IMsQuicTrace : ILogger
```
Comment (davidfowl) [inline src/Servers/Kestrel/Transport.MsQuic/src/IMsQuicTrace.cs]: internal

File: src/Servers/Kestrel/Transport.MsQuic/src/Internal/MsQuicStream.cs [hunk 31]
```diff
@@ -71,13 +72,17 @@ public MsQuicStream(MsQuicApi registration, MsQuicConnection connection, MsQuicT
             SetCallbackHandler();
 
             _processingTask = ProcessSends();
+
+            // Concatenate stream id with ConnectionId.
+            var connectionId = ConnectionId;
```
Comment (davidfowl) [inline src/Servers/Kestrel/Transport.MsQuic/src/Internal/MsQuicStream.cs]: We'll want to do this lazily.

**Review by halter73 — APPROVED** [review]

**Comment by jkotalik** [comment]
@aspnet/build many dotnet watch tests are failing on macos consistently now. Was there a downstream dependency change that happened?

**Comment by dougbu** [comment]
@jkotalik your `Skip` conditions seem extreme given the problem was specific to macOS.

@anurse I am not aware of what changed w.r.t. the macOS dotnet-watch tests but think I saw you commented about problems there in another issue. What's the scoop?

**Comment by anurse** [comment]
#16912 is tracking this.

**Comment by jkotalik** [comment]
This PR will eventually be a non-breaking change after we do work with multiplexed factories.
